source/val/validate_cfg.cpp

```diff
@@ -48,6 +48,27 @@ spv_result_t ValidatePhi(ValidationState_t& _, const Instruction* inst) {
               "basic blocks.";
   }
 
+  const Instruction* type_inst = _.FindDef(inst->type_id());
+  assert(type_inst);
+
+  const SpvOp type_opcode = type_inst->opcode();
+  if (type_opcode == SpvOpTypePointer &&
+      _.addressing_model() == SpvAddressingModelLogical) {
+    if (!_.features().variable_pointers &&
+        !_.features().variable_pointers_storage_buffer) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << "Using pointers with OpPhi requires capability "
+             << "VariablePointers or VariablePointersStorageBuffer";
+    } else if (!_.features().variable_pointers) {
```

Comment:
I can't find this restriction. The spec only limits storage class to StorageBuffer for OpLoad and OpStore. It seems to be an oversight that OpCopyMemory is not also restricted.

Reply:
In the [capabilities table](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Capability) it says

>VariablePointersStorageBuffer
Allow variable pointers, each confined to a single Block-decorated struct in the StorageBuffer storage class.

It seem that adding the `VariablePointersStorageBuffer` capability only allows certain types of variable pointers.  Says nothing about loads or stores.

However, I do see in the [validation rules](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_validation_a_validation_rules) it seems to discuss what is allowed based on loads and stores.

I also just noticed that the validation rules explicitly allow coping pointers using `OpCopyObject`, but that is defined as a [variable pointer]([url](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#VariablePointer)).

I'll need to go back of some of this.